---
title: "RNA_Seq_Alignment_Assembly"
output: html_document
---

IMPORTANT: THIS IS WRITTEN FOR RUNNING IN CHUNKS- NOT THE WHOLE DOCUMENT AT ONCE

Aligning and Assembly of RNA Seq Reads to the Genome

A. Set up your working directory
1. Create a project folder in the $HOME/RNA_Seq directory.
2. Import your data and unpack to its own folder in the project directory. (i.e. $HOME/RNA_Seq/myproject)
3. Create the following folders with their necessary files:

folder      file                    Example                   no. files       Description
/           SampleInfo.csv          geuvadis_phenodata.csv    1 file          user-generated file: Sample IDs with their phenotype, group and population characteristics
/           mergelist.txt           mergelist.txt             1 file          user-generated file: List of Sample IDs with a .gtf suffix.
/genes      AnnotatedGenome.gtf     chrX.gtf                  1 file          human/mouse gene annotations for the RefSeq database (i.e. GRCh38)
/genome     RawSequence.fa          chrX.fa                   1 file          human/mouse sequence (i.e. GRCh38 build 81)
/indexes    Genome_tran.(#).ht2     chrX_tran.(1-8).ht2       multiple        Hisat2 indexes for the relevant genome
/samples    Sample.fastq.gz         ERR###_ChrX_#.fastq.gz    multiple        raw illumina fastq files

Your working directory now looks like

/RNA_Seq
  /Project
    SampleInfo.csv
    mergelist.txt
    
    /genes
      AnnotatedGenome.gtf 
    /genome
      RawSequence.fa
    /indexes
      Genome_tran.(#).ht2
    /samples
      Sample.fastq.gz
      
Why so much effort to plan our directories? 1) the code can be copy/pasted more easily, 2) it will standardise all project data handling.

B. Read Alignment (hisat2)
This step will align your reads, it should be preceded by quality assessment and any necessary read trimming (will need to learn this and add it in)

1. Read the help file for hisat2.
nb: Pay attention to Hisat2 defaults for Phred scores, cores -p  (run 8 for multhreaded i7), file input default .fw/.fastq

```{bash}
hisat2
```

2. In our training example we will align every read with an individual command.
After this step your working directory will contain a sam file for each sample.

In the below example
-p 8        alignment will run across 8 cores (multhreading supported)
--dta       Reports alignments tailored for transcript 
-x          is followed by location and prefix of hisat2 index files
-1/-2       is followed by location and filename of reads 1 and 2
-S          is followed by a name for the output sam file (location will be in same directory as terminal)

```{bash, echo=FALSE}
cd $HOME/RNA_Seq/training
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188044_chrX_1.fastq.gz -2 chrX_data/samples/ERR188044_chrX_2.fastq.gz -S ERR188044_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188104_chrX_1.fastq.gz -2 chrX_data/samples/ERR188104_chrX_2.fastq.gz -S ERR188104_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188234_chrX_1.fastq.gz -2 chrX_data/samples/ERR188234_chrX_2.fastq.gz -S ERR188234_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188245_chrX_1.fastq.gz -2 chrX_data/samples/ERR188245_chrX_2.fastq.gz -S ERR188245_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188257_chrX_1.fastq.gz -2 chrX_data/samples/ERR188257_chrX_2.fastq.gz -S ERR188257_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188273_chrX_1.fastq.gz -2 chrX_data/samples/ERR188273_chrX_2.fastq.gz -S ERR188273_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188337_chrX_1.fastq.gz -2 chrX_data/samples/ERR188337_chrX_2.fastq.gz -S ERR188337_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188383_chrX_1.fastq.gz -2 chrX_data/samples/ERR188383_chrX_2.fastq.gz -S ERR188383_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188401_chrX_1.fastq.gz -2 chrX_data/samples/ERR188401_chrX_2.fastq.gz -S ERR188401_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188428_chrX_1.fastq.gz -2 chrX_data/samples/ERR188428_chrX_2.fastq.gz -S ERR188428_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR188454_chrX_1.fastq.gz -2 chrX_data/samples/ERR188454_chrX_2.fastq.gz -S ERR188454_chrX.sam
hisat2 -p 8 --dta -x chrX_data/indexes/chrX_tran -1 chrX_data/samples/ERR204916_chrX_1.fastq.gz -2 chrX_data/samples/ERR204916_chrX_2.fastq.gz -S ERR204916_chrX.sam
```

C. SAM Sorting and BAM Conversion (samtools)
Sorting basically arranges all your alignment result in order from left to right on the genome
Conversion prepares your sam file output by hisat2 into a bam file format for use with StringTie

1. Read the help file for samtools.

```{bash}
samtools
```

2. In our training example we will sort and convert every sam file with an individual command. 
After this step your working directory will contain a bam file for each sample.

In the below example
sort        Sorts alignments
-@ 8        Number of BAM compression threads to use in addition to main thread [0]. 
-o          location and filename for output bam file

```{bash}
cd $HOME/RNA_Seq/training
samtools sort -@ 8 -o ERR188044_chrX.bam ERR188044_chrX.sam
samtools sort -@ 8 -o ERR188104_chrX.bam ERR188104_chrX.sam
samtools sort -@ 8 -o ERR188234_chrX.bam ERR188234_chrX.sam
samtools sort -@ 8 -o ERR188245_chrX.bam ERR188245_chrX.sam
samtools sort -@ 8 -o ERR188257_chrX.bam ERR188257_chrX.sam
samtools sort -@ 8 -o ERR188273_chrX.bam ERR188273_chrX.sam
samtools sort -@ 8 -o ERR188337_chrX.bam ERR188337_chrX.sam
samtools sort -@ 8 -o ERR188383_chrX.bam ERR188383_chrX.sam
samtools sort -@ 8 -o ERR188401_chrX.bam ERR188401_chrX.sam
samtools sort -@ 8 -o ERR188428_chrX.bam ERR188428_chrX.sam
samtools sort -@ 8 -o ERR188454_chrX.bam ERR188454_chrX.sam
samtools sort -@ 8 -o ERR204916_chrX.bam ERR204916_chrX.sam
```

***** I NEED TO VERIFY THE PIPELINE FOR USING LIMMA *****

D. Assembly and Quantification (stringtie)
1. Read the help file for stringtie

```{bash}
stringtie
```

2. In our training example we will sort and convert every bam file with an individual command. 
After this step your working directory will contain a gtf text document for each sample.

```{bash}
cd $HOME/RNA_Seq/training
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188044_chrX.gtf -l ERR188044 ERR188044_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188104_chrX.gtf -l ERR188104 ERR188104_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188234_chrX.gtf -l ERR188234 ERR188234_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188245_chrX.gtf -l ERR188245 ERR188245_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188257_chrX.gtf -l ERR188257 ERR188257_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188273_chrX.gtf -l ERR188273 ERR188273_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188337_chrX.gtf -l ERR188337 ERR188337_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188383_chrX.gtf -l ERR188383 ERR188383_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188401_chrX.gtf -l ERR188401 ERR188401_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188428_chrX.gtf -l ERR188428 ERR188428_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR188454_chrX.gtf -l ERR188454 ERR188454_chrX.bam
stringtie -p 8 -G chrX_data/genes/chrX.gtf -o ERR204916_chrX.gtf -l ERR204916 ERR204916_chrX.bam
```

3. Merge Stringtie results
Here we will take all the sample gtf data and merge it into a single file called stringtie_merged.gtf

```{bash, echo=FALSE}
cd $HOME/RNA_Seq/training
stringtie --merge -p 8 -G chrX_data/genes/chrX.gtf -o stringtie_merged.gtf chrX_data/mergelist.txt
```

E. Compare Transcripts with Reference Annotation (gffcompare)
1. Read the help file for gffcompare

```{bash}
gffcompare
```

2. In our training example we will compare the stringtie results with the reference annotation. 
All six files will have the prefix 'merged.' and be added to the working directory.

```{bash}
cd $HOME/RNA_Seq/training
gffcompare -r chrX_data/genes/chrX.gtf -G -o merged stringtie_merged.gtf
```

F. Transcript Abundance and Table Counts (stringtie)
1. In our training example we will estimate transcript abundances and create the table count files for Ballgown
After this step the directory /ballgown will appear in your working directory with structure (where ERR188044 is the sample id)

/ballgown
  /ERR188044
    e2t.ctab
    e_data.ctab
    ERR188044_chrX.gtf
    i2t.ctab
    i_data.ctab
    t_data.ctab

The new gtf file is now in a compatibale format for ballgown. 
No idea what all the other files are for- probably just procedurally generated and read by the program to avoid holding everything in memory.

```{bash}
cd $HOME/RNA_Seq/training
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188044/ERR188044_chrX.gtf ERR188044_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188104/ERR188104_chrX.gtf ERR188104_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188234/ERR188234_chrX.gtf ERR188234_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188245/ERR188245_chrX.gtf ERR188245_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188257/ERR188257_chrX.gtf ERR188257_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188273/ERR188273_chrX.gtf ERR188273_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188337/ERR188337_chrX.gtf ERR188337_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188383/ERR188383_chrX.gtf ERR188383_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188401/ERR188401_chrX.gtf ERR188401_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188428/ERR188428_chrX.gtf ERR188428_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR188454/ERR188454_chrX.gtf ERR188454_chrX.bam
stringtie -e -B -p 8 -G stringtie_merged.gtf -o ballgown/ERR204916/ERR204916_chrX.gtf ERR204916_chrX.bam
```

