---
title: "RNA_Seq"
output: html_notebook
---

IMPORTANT: WHY CAN'T I JUST RUN THIS FROM R NOTEBOOKS?
ANSWER: BECAUSE IF YOU ARE FOLLOWING THIS YOU DON'T HAVE R NOTEBOOKS. COPY-PASTE FROM THE MARKDOWN (ALTHOUGH THE FORMATTING WILL BE SHIT)

A. Install Linux Mint on Windows Machine (dual boot)
nb: partitioning can also be done using GParted within a live boot linux

1. Install Linux Mint XFCE 18.1 “Serena” 64bit
2. Set extended partition root /. Do not specify Home partition: it will share the same parition with root as we will use an NTFS partition to share data between windows and linux.
3. Set swap space: I have used 60Gb
4. Allow 10% ssd unallocated for over provisioning

B. Configure synaptic to consider package dependencies
1. Settings > Preferences > Consider recommended packages as dependencies

C. Install Anaconda
nb. Anaconda is a python data science distribution. Even if not using python it will aid the installation of R pacakages.

1. Download Anaconda (https://docs.continuum.io/anaconda/install) to target directory /home/jeff
2. From terminal in same directory



```{bash}
bash ~/Downloads/Anaconda3-4.3.0-Linux-x86_64.sh
```

3. Choose yes to prepend PATH- this will allow you to call conda command from shell

D. Install Bioconda
1. If you’ve installed anaconda you already have bioconda. nb. bioconda supports only 64-bit Linux and Mac OSX.

E. Install Biopython
1. From any terminal

```{bash}
conda install -c anaconda biopython=1.68 
```

F. Install R-base and R-base-dev
nb. Do not use synaptic install- repository packages may be out of date
nb. The procedure for installing these packages may be affected by your Linux distribution
1. Linux Mint 18 “Serena” is based on Ubuntu 16.04 LTS “Xenius Xenial”. Install guide here: https://www.datascienceriot.com/how-to-install-r-in-linux-ubuntu-16-04-xenial-xerus/kris/
2. Add R to apt-get repository sources

```{bash}
sudo echo "deb http://cran.rstudio.com/bin/linux/ubuntu xenial/" | sudo tee -a /etc/apt/sources.list
```

3. Add R to Ubuntu Keyring. From any terminal

```{bash}
 gpg --keyserver keyserver.ubuntu.com --recv-key E084DAB9
 gpg -a --export E084DAB9 | sudo apt-key add -
```

4. Install R-Base. From any terminal

```{bash}
sudo apt-get update
sudo apt-get install r-base r-base-dev
```

5. Install R-Studio. From any terminal

```{bash}
sudo apt-get install gdebi-core
wget https://download1.rstudio.org/rstudio-1.0.44-amd64.deb
sudo gdebi -n rstudio-1.0.44-amd64.deb
rm rstudio-1.0.44-amd64.deb
```

G. Install irkernel for use in Jupyter notebooks. Download https://anaconda.org/r/r-irkernel
1. From any terminal

```{bash}
conda install -c r r-irkernel=0.7.1
```

H. Read the Hisat2 RNA-Seq protocol
1. Pertea et al, Nat Prot 2016, Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown

I. Create your working directories for RNA-Seq Analysis
nb. can also create folders in the file browser.
nb. the $HOME syntax points to /home/user (i.e. /home/jeff)

```{bash}
mkdir $HOME/RNA_Seq
mkdir $HOME/RNA_Seq/tools
mkdir $HOME/bin
```

J. Add bin to PATH
Export the excutables folder /bin to PATH. Verify the export by echo.

```{bash}
export PATH=$HOME/bin:$PATH
echo $PATH
```

K. Install SAM tools
1. Download the source tar file from http://www.htslib.org/download/
2. Copy tar file to $HOME/RNA_Seq/tools
3. From a terminal in the same directory

```{bash}
tar jxvf samtools-1.3.1.tar.bz2
```

4. Build tar file from source. From a terminal in the unpacked folder $HOME/RNA_Seq/tools/samtools-1.3.1

```{bash}
make
```

5. Copy samtools binary into PATH. From a terminal in the parent /tools directory
nb. As the bin $HOME/bin directoryhas already been exported to PATH it is not necessary to keep doing so afte new executables have been added

```{bash}
cp samtools-1.3.1/samtools $HOME/bin 
```

L. Install Hisat2
1. Download the Hisat2 linux binary to RNA_Seq/tools. http://ccb.jhu.edu/software/hisat2/index.shtml

```{bash}
unzip hisat2-2.0.5-Linux_x86_64.zip
```

2. Copy the hisat executables into $HOME/bin. From a terminal in the unzipped directory.

```{bash}
cp hisat2-2.0.5/hisat2* hisat2-2.0.5/*.py $HOME/bin
```

M. Install StringTie
1. Download the StringTie linux binary to RNA_Seq/tools. http://ccb.jhu.edu/software/stringtie/
2. Unpack the tar file from the same directory

```{bash}
tar xvzf stringtie-1.3.3.Linux_x86_64.tar.gz
```

3. Copy StringTie executable to $HOME/bin. From a terminal in the unpacked directory.

```{bash}
cp stringtie-1.3.3.Linux_x86_64/stringtie /home/jeff/RNA_Seq/tools
```

N. Install git
nb. git tools are required to fetch the gffcompare package. From any terminal.

```{bash}
sudo apt install git
```

O. Install gffcompare and its GCLib dependency
1. From a terminal in RNA_Seq/tools

```{bash}
git clone https://github.com/gpertea/gclib
git clone https://github.com/gpertea/gffcompare
cd gffcompare
make release
```

nb. If you get something like this:

fatal: unable to access 'https://github.com/gpertea/gclib/': Could not resolve host: github.com

then you have an interfering proxy setting. So from any terminal:

```{bash}
git config --global --unset https.proxy (or http if git url says http)
```

and then re-enter the commands at O.1.

P. Install Bioconductor
nb. The Nature protocol requires Bioconductor version 3.0 or greater and R version 3.1
1. Open an R Terminal

```{r}
install.packages("devtools",repos="http://cran.us.r-project.org")
```

nb. if you get any errors or missing dependencies its probably because of missing dependencies and lack of permissions as R user, therefore:

```{bash}
sudo apt-get -y install libcurl4-gnutls-dev libxml2-dev libssl-dev
sudo su
R
```

then in the open R terminal (now as root)

```{r}
install.packages('devtools', repos='http://cran.rstudio.com/')
source("https://www.bioconductor.org/biocLite.R")
```

nb. Use http if https doesnt work.

Q. Install the necessary Bioconductor packages
nb. This command fetches a script from a private repository to install ballgown, genefilter, dplyr, devtools.
nb. You'll get a shitfest of console output- watch it for errors.

```{r}
biocLite(c("alyssafrazee/RSkittleBrewer","ballgown","genefilter","dplyr","devtools"))
```

R. Setup tools for data preprocessing- 
FastQ dependencies
Fast Q screen http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/
Fast QC http://www.bioinformatics.babraham.ac.uk/projects/fastqc/

